addons/stock/static/src/stock_forecasted/forecasted_details.js

```diff
@@ -44,31 +46,158 @@ export class ForecastedDetails extends Component {
+                if (rule.condition(line)) {
+                    const groupKey = rule.key(line);
+                    (rule.target[groupKey] ??= []).push(line);
+                }
+            }
+        }
+    }
+
+    _prepareLines(){
+        if (this.multipleProducts) {
+            this.props.docs.lines.sort((a, b) => (a.product.id || 0) - (b.product.id || 0));
+        }
+    }
+
+    _prepareData(){
+        let lines = this.lines;
+        this.onHandCondition = lines.length && !lines.some((line) => line.document_in || line.replenishment_filled);
+
+        this.OnHandTotalQty = Object.fromEntries(
+            Object.entries(this.OnHandLinesPerProduct).map(([id, lines]) => [
+                id,
+                lines.reduce((sum, line) => sum + line.quantity, 0),
+            ])
+        );
+        this.AvailableOnHandTotalQty = Object.fromEntries(
+            Object.entries(this.OnHandLinesPerProduct).map(([id, lines]) => [
+                id,
+                lines.reduce((sum, line) => sum + (line.reservation ? 0 : line.quantity), 0),
+            ])
+        );
+        for (const productId of this.productIds){
+            if (!(productId in this.FreeStockLinesPerProduct) || !(productId in this.LinesPerProduct)){
+                continue;
+            }
+            const lines = this.FreeStockLinesPerProduct[productId]
+            if (this.LinesPerProduct[productId].length > 1 && lines.length == 1 && lines[0]?.quantity === 0 ){
+                const removeIndex = this.lines.indexOf(lines[0]);
+                this.lines.splice(removeIndex,1);
+            }
+        }
+    }
+
+    _mergeLines(){
+        let lines = this.lines;
+        this.mergesLinesData = {};
+        let j = 0;
+        for(let i = 0; i < lines.length-1; i++){
+            const index = i-j;
+            const line = lines[index];
+            const nextLine = lines[i + 1];
+            if (line.product.id != nextLine.product.id || !this._sameLineRule(line, nextLine)) {
+                j = 0;
+                continue;
+            }
+            this.mergesLinesData[index] = this.mergesLinesData[index] || {
+                rowcount: 1,
+                tot_qty: line.quantity,
+            };
+            this.mergesLinesData[index].rowcount += 1;
+            this.mergesLinesData[index].tot_qty += nextLine.quantity;
+            j++;
+        }
+    }
+
+    _sameLineRule(line, nextLine){
+        const OnHand = this.OnHandLinesPerProduct[line.product.id] || [];
+        return  this.sameDocumentIn(line, nextLine) || (OnHand.includes(line) && OnHand.includes(nextLine))
+    }
+
     displayReserve(line){
-        return this.props.docs.user_can_edit_pickings && !line.in_transit && this.canReserveOperation(line);
+        let isOnHandSplittedLine = this.OnHandLinesPerProduct[line.product.id] && !this.isOnHand(this.line_index) && this.OnHandLinesPerProduct[line.product.id].some(l => this.sameDocumentOut(l, line))
```

Comment:
The condition in backend to reserve. Is to have free stock. I guess you could reserve everything that is not reserve if you have free stock
(and maybe only if the line hasn't been split